File: 133_CloneGraph.cpp
// Topic: Clone Graph
// Description: Given a reference of a node in a connected undirected graph, return a deep copy
// (clone) of the graph. Approach: Read an adjacency list, build nodes, then BFS clone using hashmap
// from original->copy.

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int val;
    vector<Node*> neighbors;
    Node(int v) : val(v) {}
};

Node* buildGraph(int n, const vector<vector<int>>& adj)
{
    if (n == 0)
        return nullptr;
    vector<Node*> nodes(n + 1, nullptr);
    for (int i = 1; i <= n; ++i) nodes[i] = new Node(i);
    for (int i = 1; i <= n; ++i)
    {
        for (int v : adj[i]) nodes[i]->neighbors.push_back(nodes[v]);
    }
    return nodes[1];  // return node 1 as entry
}

Node* cloneGraph(Node* node)
{
    if (!node)
        return nullptr;
    unordered_map<Node*, Node*> mp;
    queue<Node*> q;
    q.push(node);
    mp[node] = new Node(node->val);
    while (!q.empty())
    {
        Node* cur = q.front();
        q.pop();
        for (Node* nei : cur->neighbors)
        {
            if (!mp.count(nei))
            {
                mp[nei] = new Node(nei->val);
                q.push(nei);
            }
            mp[cur]->neighbors.push_back(mp[nei]);
        }
    }
    return mp[node];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<vector<int>> adj(n + 1);
    for (int i = 1; i <= n; ++i)
    {
        int k;
        cin >> k;
        adj[i].reserve(k);
        for (int j = 0; j < k; ++j)
        {
            int v;
            cin >> v;
            adj[i].push_back(v);
        }
    }
    Node* start = buildGraph(n, adj);
    Node* cloned = cloneGraph(start);
    // Output cloned adjacency in same format (BFS to ensure we list each node once)
    unordered_map<int, vector<int>> out;
    unordered_set<int> seen;
    queue<Node*> q;
    q.push(cloned);
    seen.insert(cloned->val);
    while (!q.empty())
    {
        Node* cur = q.front();
        q.pop();
        for (Node* nei : cur->neighbors)
        {
            out[cur->val].push_back(nei->val);
            if (!seen.count(nei->val))
            {
                seen.insert(nei->val);
                q.push(nei);
            }
        }
    }
    // print n lines; if isolated nodes exist their adjacency will be empty
    for (int i = 1; i <= n; ++i)
    {
        auto& vec = out[i];
        cout << vec.size();
        for (int v : vec) cout << " " << v;
        cout << "\n";
    }
}

/*
Time Complexity: O(N + E)
Space Complexity: O(N)
*/

File: 133_CloneGraph.py
# Topic: Clone Graph
# Description: Given a reference of a node in a connected undirected graph, return a deep copy (clone) of the graph.
# Approach: BFS clone using hashmap from original->copy.

from typing import Optional, Dict, List
from collections import deque


class Node:
    def __init__(self, val: int = 0, neighbors: List['Node'] = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
        if not node:
            return None
        cloned: Dict[Node, Node] = {}
        queue = deque([node])
        cloned[node] = Node(node.val)
        while queue:
            cur = queue.popleft()
            for nei in cur.neighbors:
                if nei not in cloned:
                    cloned[nei] = Node(nei.val)
                    queue.append(nei)
                cloned[cur].neighbors.append(cloned[nei])
        return cloned[node]


def build_graph(adj_list: List[List[int]]) -> Optional[Node]:
    if not adj_list:
        return None
    n = len(adj_list)
    nodes = [Node(i + 1) for i in range(n)]
    for i, neighbors in enumerate(adj_list):
        for j in neighbors:
            nodes[i].neighbors.append(nodes[j - 1])
    return nodes[0]


if __name__ == "__main__":
    tests = [[[2, 4], [1, 3], [2, 4], [1, 3]], [[]], []]
    for i, adj in enumerate(tests):
        original = build_graph(adj) if adj else None
        cloned = Solution().cloneGraph(original)
        print(f"Test {i + 1}: {'Cloned' if cloned else 'None'}")

"""
Time Complexity: O(N + E)
Space Complexity: O(N)
"""
